<html>
    <head>
        <script type="text/javascript" src="./site_media/jquery/js/jquery-1.6.1.min.js"></script>
        <script type="text/javascript" src="./site_media/jquery/js/jquery-ui-1.8.13.min.js"></script>
		<script type="text/javascript" src="./site_media/jquery/js/tipTipX/jquery.tipTipX.js"></script>
        <link rel="stylesheet" type="text/css" href="./site_media/stylesheet.css"/>
        <link type="text/css" href="./site_media/jquery/css/aristo/jquery-ui-1.8.7.custom.css" rel="stylesheet" />
        <link href="./site_media/jquery/js/tipTipX/jquery.tipTipX.css" rel="stylesheet" type="text/css" />
        <link rel="stylesheet" type="text/css" href="./site_media/report.css" media="screen" />
		<script type="text/javascript">
			$(function() {
				$().tipTipDefaults({ delay : 0 });
				$(".tip").tipTip({ position : "bottom" });
			});
		</script>
		<script type="text/javascript">
			$(function() {
				$( "#accordion").accordion();
			});
		</script>
</head>
 <body>
        <div class="all all_width">
            <div class="topbar">
                <div class="logoholder">
                    <a href="http://www.edgebio.com/"><img src="./site_media/images/raw_name_small.png" alt="EdgeBio Systems" style="border:none;"/></a>
                </div>
            </div>
        </div>
  <div style="width:780px;margin-left:auto;margin-right:auto;height:100%">
   <h1><center><span style="cursor:help" title="Targetted regions='/ngs/reference data/genomes/Hsapiens/staging/Ov SS bed files/SureSelect/SS regions CP'   Target padding=500">Target Coverage Analysis Report</span></center></h1>
<h3><center>OvationFFPE-ready</center></h3>
<style type="text/css">
  table {width:100% !important;border-collapse:collapse;margin:0;table-layout:fixed}
  th,td {font-family:"Lucida Sans Unicode","Lucida Grande",Sans-Serif;font-size:14px;line-height:1.2em;font-weight:normal}
  th,td {width:50% !important;border:1px solid #bbbbbb;padding:5px;text-align:center}
  td {padding-top:20px;padding-bottom:20px}
  td.inleft  {width:75% !important;border-width:0;text-align:left;padding:2px;padding-left:40px}
  td.inright {width:25% !important;border-width:0;text-align:right;padding:2px;padding-right:40px}
  img.frm {display:block;margin-left:auto;margin-right:auto;margin-top:10px;margin-bottom:10px;width:700;height:700;border-width:0;cursor:help}
  .thelp {cursor:help}
</style>
<center><table>
<tr><th><span class="thelp" title="All reads mapped to the reference. ">All Reads</span></th></tr>
<tr><td><table>
<tr><td class="inleft"><span class="thelp" title="Total number of reads mapped to the reference. ">Number of mapped reads</span></td> <td class="inright">1,625,036</td></tr>
<tr><td class="inleft"><span class="thelp" title="Total number of reads mapped to any targeted region of the reference. A read is considered to be on target if at least one aligned base overlaps a target region. A read that overlaps a targeted region but where only flanking sequence is aligned, e.g. due to poor matching of 5' bases of the read, is not counted. ">Number of reads on target</span></td> <td class="inright">1,021,939</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of reads mapped to any targeted region relative to all reads mapped to the reference. ">Percent reads on target</span></td> <td class="inright">62.89%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of reads mapped to any padded targeted region relative to all reads mapped to the reference. This typically includes reads of probe-enriched DNA fragments that overlapped a target region but were only sequenced over bases flanking that region. ">Percent reads on padded target</span></td> <td class="inright">89.93%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The total number of bases covered by reads aligned to the reference. ">Total aligned base reads</span></td> <td class="inright">61,749,848</td></tr>
<tr><td class="inleft"><span class="thelp" title="The total number of target bases covered by any number of aligned reads.  ">Total base reads on target</span></td> <td class="inright">33,727,151</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percent of all bases covered by reads aligned to the reference that covered bases in target regions. ">Percent base reads on target</span></td> <td class="inright">54.62%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The total number of bases in all target regions of the reference. ">Bases in targeted reference</span></td> <td class="inright">1,140,710</td></tr>
<tr><td class="inleft"><span class="thelp" title="The total number of target bases that had at least one read aligned over the proximal sequence. Only the aligned parts of each read are considered, e.g. un-aligned (soft-cut) bases at the 5' ends of mapped reads are not considered. Covered reference target bases may include sample DNA read base mismatches but does not include read base deletions in the read nor insertions between reference bases. ">Bases covered (at least 1x)</span></td> <td class="inright">1,052,115</td></tr>
<tr><td class="inleft"><span class="thelp" title="The average number of reads of all targeted reference bases. This is the total number of base reads on target divided by the number of targeted bases and therefore includes any bases that had no coverage. ">Average base coverage depth</span></td> <td class="inright">29.57</td></tr>
<tr><td class="inleft"><span class="thelp" title="The maximum number of times any single target base was read. ">Maximum base read depth</span></td> <td class="inright">307</td></tr>
<tr><td class="inleft"><span class="thelp" title="The average number of reads of all targeted reference bases that were read at least once. ">Average base read depth</span></td> <td class="inright">32.06</td></tr>
<tr><td class="inleft"><span class="thelp" title="The standard deviation (root-variance) of the read depths of all targeted reference bases that were read at least once. ">Std.Dev base read depth</span></td> <td class="inright">27.31</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of target bases covered by at least one read. ">Target coverage at 1x</span></td> <td class="inright">92.233%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of target bases covered by at least five reads. ">Target coverage at 5x</span></td> <td class="inright">83.661%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of target bases covered by at least ten reads. ">Target coverage at 10x</span></td> <td class="inright">72.896%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of target bases covered by at least twenty reads. ">Target coverage at 20x</span></td> <td class="inright">54.553%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of target bases covered by at least fifty reads. ">Target coverage at 50x</span></td> <td class="inright">19.890%</td></tr>
<tr><td class="inleft"><span class="thelp" title="The percentage of target bases covered by at least one hundred reads. ">Target coverage at 100x</span></td> <td class="inright">2.552%</td></tr>
<tr><td class="inleft"><span class="thelp" title="Target coverage at 500x">Target coverage at 500x</span></td> <td class="inright">0.000%</td></tr>
<tr><td class="inleft"><span class="thelp" title="Target coverage at 1000x">Target coverage at 1000x</span></td> <td class="inright">0.000%</td></tr>
<tr><td class="inleft"><span class="thelp" title="Target coverage at 2500x">Target coverage at 2500x</span></td> <td class="inright">0.000%</td></tr>
<tr><td class="inleft"><span class="thelp" title="Target coverage at 5000x">Target coverage at 5000x</span></td> <td class="inright">0.000%</td></tr>
<tr><td class="inleft"><span class="thelp" title="Target coverage at 10000x">Target coverage at 10000x</span></td> <td class="inright">0.000%</td></tr>
</table></td></tr>
<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage.png" title="Target base coverage is plotted against read depth, where read depth is the number of times a particular base is read and coverage is the number of counts of bases read at that read depth. This plot includes the number of target bases that were not read, i.e to 0x coverage. The right hand y-axis shows the cumulative coverage as a percentage of the total number of target base reads. This may be used to gauge to percentage of target bases read to a particular depth. ">Target Coverage<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage.png" alt="Target Coverage"/></a> <a href="all_reads/OvationFFPE-ready.coverage.xls">Download data file</a></td>
</tr>

<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage_binned.png" title="Target base coverage is plotted against read depth, where read depth is the number of times a particular base is read and coverage is the number of counts of bases read at that read depth. The read depths are binned to better represent coverage where the maximum read depth is high and the binning size is chosen accordingly. Hence an x-axis value of 20x indicates the sum of coverage for read depths from 1 to 20, or 11 to 20 if the preceding bar was labelled 10x, etc. This plot does not include the coverage for non-covered target bases. The right hand y-axis shows the cumulative coverage as a percentage of the total number of target base reads. ">Binned Target Coverage<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage_binned.png" alt="Binned Target Coverage"/></a> <a href="all_reads/OvationFFPE-ready.coverage_binned.xls">Download data file</a></td>
</tr>

<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage_onoff_target.png" title="The number of reads that align to each chromosome of the reference are plotted as a bar chart. The number of reads that have aligned sequence overlapping any target region are represented by the gray portion of the plot and those aligned off-target are represented by the white region. ">Target Coverage by Chromosome<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage_onoff_target.png" alt="Target Coverage by Chromosome"/></a> <a href="all_reads/OvationFFPE-ready.coverage_by_chrom.xls">Download data file</a></td>
</tr>

<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage_onoff_padded_target.png" title="The number of reads that align to each chromosome of the reference are plotted as a bar chart. The number of reads that have aligned sequence overlapping any target region are represented by the gray portion of the plot and those aligned off-target are represented by the white region. Here the individual target regions are extended (padded) on both ends, e.g. by 100 bases. Hence, the on-target counts includes most reads of probe-enriched (pulled down) DNA fragments that did not extend into the adjacent target region. ">Padded Target Coverage by Chromosome<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage_onoff_padded_target.png" alt="Padded Target Coverage by Chromosome"/></a> <a href="all_reads/OvationFFPE-ready.coverage_by_chrom_padded_target.xls">Download data file</a></td>
</tr>

<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage_on_target.png" title="The number of aligned base reads are counted for individual target regions of the reference and normalized by dividing by the length of the target. These values are plotted on a bar chart for each chromosome of the reference to a common scale, set by the highest normalized count. Red areas of the bars show the fraction of the target bases uncovered by any read, e.g. 20:80 red:gray indicates only 80% of the individual target was read. Note that with large numbers of targets, details for individual target regions (bars) may not be visible. Download the data file using the link provided to examine the individual target coverage in full detail. ">Individual Target Coverage<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage_on_target.png" alt="Individual Target Coverage"/></a> <a href="all_reads/OvationFFPE-ready.fine_coverage.xls">Download data file</a></td>
</tr>

<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage_map_onoff_target.png" title="Aligned read starts are counted for each 100 bases of the reference and plotted as a bar only if the count is at least 5. Hence zero or low coverage regions, including isolated read mappings, are not represented in these plots. Plot color is alternated to show continuous regions of coverage. Red and blue represent the reads starting in a 100 base region overlapping a target region. Black and gray represent reads starting outside of a target region. Peaks are contiguous and aligned to 100 base counts along the reference but no distance between any two peaks is depicted. Note that with large numbers of targets, peak shapes and resolution between on- and off-target peaks may not be discernable. Download the data file using the link provided to examine coverage in full detail. ">On/Off Target Read Alignment<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage_map_onoff_target.png" alt="On/Off Target Read Alignment"/></a> <a href="all_reads/OvationFFPE-ready.coverage_map_by_chrom.xls">Download data file</a></td>
</tr>

<tr>
<td><a style="cursor:help" href="all_reads/OvationFFPE-ready.coverage_normalized.png" title="The fraction of total target bases covered is plotted against normalized coverage, where normalized coverage is the number of times a particular base is read (read depth) divided by the mead read depth of all target bases. The y-axis is the cumulative read depth divided by the total number of target base reads. ">Normalized Target Coverage<br/><img class="frm" src="all_reads/OvationFFPE-ready.coverage_normalized.png" alt="Normalized Target Coverage"/></a> <a href="all_reads/OvationFFPE-ready.coverage.xls">Download data file</a></td>
</tr>

</table></center>
            <!-- zebra stripe the tables -->
			<script type="text/javascript">
				 
			 $(document).ready(function(){

			   $('h2').append('<a href="javascript:;" class="expandCollapseButton" title="Collapse Section"></a>');

				$('.expandCollapseButton').toggle(function() {
					if ( $(this).attr('title') == 'Collapse Section'){
						$(this).css('background-position','right top');
						$(this).attr('title','Expand Section');
					}else{
						$(this).css('background-position','left top');
						$(this).attr('title','Collapse Section');
					}
				}, function() {
					if ( $(this).attr('title') == 'Expand Section'){
						$(this).css('background-position','left top');
						$(this).attr('title','Collapse Section');
					}else{
						$(this).css('background-position','right top');
						$(this).attr('title','Expand Section');
					}
				});
				
				$('.expandCollapseButton').click(function(event){
					$(event.target).parent().parent().toggleClass('small');
					$(event.target).parent().next().slideToggle();
				});

				$('#tf .expandCollapseButton').css('background-position','right top');$('#tf .expandCollapseButton').attr('title','Expand Section');$('#tf').parent().toggleClass('small');$('#tf').next().toggle();
				//start overlay

				$(".heading tbody tr").mouseover(
						function(){
							$(this).addClass("table_hover");
							
				}).mouseout(
						function(){
							$(this).removeClass("table_hover");
				});

				$(".noheading tbody tr").mouseover(
						function(){
							$(this).addClass("table_hover");
							
				}).mouseout(
						function(){
							$(this).removeClass("table_hover");
				});
				
				$(".heading tr:odd").addClass("zebra");
				$(".noheading tr:odd").addClass("zebra");

				$('#datatable').dataTable( {
					"aLengthMenu": [[0], [0]],
					"sDom": '<"top"fl>rt<"bottom"ip><"spacer">', 
					"sScrollX": "100%",
					"sScrollXInner": "110%",
					"bScrollCollapse": true,
					"sPaginationType": "full_numbers"
				} );		
			});
			 
		  </script>


  <div class="clear"></div>
  <div class="footer">
    <a href=http://www.edgebio.com/techinfo/techinfosupport.php>Request Support</a>
  </div>
  <br/><br/></div>
 </body></html>
